gistmgr.1
\"
\" This manual is under the GNU Free Documentation License 1.3.
\" A copy of the license should be in DOCS-LICENSE, if not see
\" the second url under the COPYING section.
\"
.TH "GISTMGR" "1" "December 2017" "GistMGR" "GistMGR Manual"
.SH "NAME"
\fBgistmgr\fR \- manages URLs compatible with GitHub's Gists

.SH "SYNOPSIS"
.IP "Multi-List Files (default)"
.RS
gistmgr \fBadd \fIlist url \fR[\fIcomment\fR]

gistmgr \fBremove \fIlist url\fR

gistmgr \fBlist \fIlist\fR
.RE
.IP "Single-List Files"
.RS
gistmgr \fBadd \fIurl \fR[\fIcomment\fR]

gistmgr \fBremove \fIurl\fR

gistmgr \fBlist\fR
.RE
.SH "DESCRIPTION"
GistMGR is a
.BR sh (1)
script that uses
.BR grep (1)
and standard
.BR sh (1)
in/output redirectors through
.BR git (1)
to provide a seamless experience for managing URLs
inside of a text file denoted in
.IR ~/.gistmgrrc ,
by default
.IR ~/.gistmgr/gistlist .
It provides three commands to manage files: \fB
.BR add ,
.BR remove ", and"
.BR list .

\fBadd\fR adds a new
.IR url
to the list
.IR list
with comment
.IR comment ,
if applicable.
If urlfile is a single-list file, then it adds
.IR url
to the list with comment
.IR comment "if applicable."
It appends the line to
.IR gistlist ,
and runs
.BR git-commit (1)
to log the change into the Git repository.

\fBremove\fR removes
.IR url
from the list
.IR list
with an inverse
.BR grep (1)
which removes
.BR ALL
instances of the URL from
.IR gistlist ,
and then runs
.BR git-commit (1)
to log the change into the Git repository.
If urlfile is a single-list file, then it removes
.IR url
from the file directly.

\fBlist\fR lists all the URLs in 
.IR list
using a
.BR grep (1)
command. Because of how it works internally,
it allows all notation permitted by
.BR egrep (1).
It is not formatted; use
.BR gistmgrfmt (1)
for formatting. Note that on single-list files,
it accepts no arguments (but doesn't complain
if you specify more.)

.SH "SINGLE-LIST URL FILES"
.IR gistlist
has support for "Single-List URL Files". These are
files that have no concepts of 'lists'. Consider
the following urlfile:

.EX
# GistMGR
00 https://gnu.org/ GNU Project
.EE

This is good for personal lists; i.e. shopping lists,
etc. where they all can or need to be in the same
file; but not for sharing lists. In shared lists, the
list usually has a single purpose, and there's no
point in repeating the list identifier when it won't
change. As such, on December 16th 2017 support was
added for "Single-List Files". Assuming the second
line of the file is '\fC#@#\fR', the list is interpreted
as a single-file list and the single-file commands
apply. This urlfile would look as follows:

.EX
# GistMGR
#@#
https://gnu.org/ GNU Project
.EE

Note that there is no automated way to make single-list
files; this will likely not be patched. See
.BR BUGS .
.SH "FILES"
.I ~/.gistmgrrc
.RS
The user configuration file for
.IR gistmgr .
It is sourced by
.BR sh (1),
so arbitrary commands can be put in there, as well as
comments as supported by
.BR sh (1).
.RE
.I ~/.gistmgr/gistlist
.RS
The default file where URLs are stored. They are
stored in the file in the following format:

.IR list
.IR url
comments...
.RE
.SH "EXIT STATUS"
If all goes well, exit status is 0.

If no URLs are found in
.BR list
query, exit status is 1.

If an environment variable that is required is not
set in
.IR ~/.gistmgrrc ,
then exit status is 2.

If there is an invalid number of arguments, then
exit status is 3.

.SH "ENVIRONMENT"
Note that aliases of these can be set in
.IR ~/.gistmgrrc .
In aliases, remove the GISTMGR_ and make it
lowercase. For example, GISTMGR_URLFILE here
is urlfile in
.IR ~/.gistmgrrc .

.IP GISTMGR_URLFILE
.RS
Where the URLs are stored.
.RE

.SH "BUGS"
There is no easy way to create single-list URL
files, as there is multi-list.

PGP encrypting through
.BR gpg (1)
is not supported in the slightest.

The code is a mess.

.SH "COPYING"
GistMGR is under the GNU General Public
License, 3.0 with NO LIABILITY OR
WARRANTY.

This manual page is under the GNU
Free Documentation License, 1.3.

A copy of both licenses should have
been included with this program,
see https://gnu.org/licenses/gpl-3.0.txt
for the GPL and https://gnu.org/licenses/fdl-1.3.txt
for the FDL if there were none included.

.SH "DEPENDENCIES"
.BR sh (1),
.BR grep (1),
.BR sed (1),
.BR mv (1)

.SH "AUTHOR"
thatlittlegit <wapidstyle@live.ca>

.SH "SEE ALSO"
.BR grep (1),
.BR git (1)


gistmgr.sh
#!/bin/sh
#
# GistMGR manages URL-based lists that are integrated with Git.
#
# GistMGR is under the GNU General Public License, 3.0.
# it is distributed in the hope that it can be useful,
# but with NO WARRANTY OR LIABILITY.
#
# A copy of the license should have been included with
# this software; if not, see http://gnu.org/licenses/gpl-3.0.txt
#
genrcfile() {
    # We assume we need to make a .gistmgrrc file, and thus won't
    # check for it.
    echo "$0: no gistmgrrc file exists, creating default" 1>&2
    echo "$0: creating local git repo, you'll need to connect it to gist" 1>&2
    mkdir "$HOME/.gistmgr" && cd "$HOME/.gistmgr"
    git init
    echo "# GistMGR" > gistlist
    git add gistlist && git commit -m "[gistmgr] Initial Commit"
    echo "urlfile=$HOME/.gistmgr/gistlist" >>~/.gistmgrrc
}

usage() {
       echo "GistMGR - by thatlittlegit

$0 add <list> <url> [comment]
$0 remove <list> <url>
$0 list <list>

GistMGR is under the GPL 3 License."
    exit 3
}

[ -f ~/.gistmgrrc ] || genrcfile

if [ $# -eq 0 ] || [ "$1" = "--help" ] || [ "$1" = "-h" ] || [ -z "$1" ]
then
    usage
else
    . ~/.gistmgrrc

    [ -z "$GISTMGR_URLFILE" ] || urlfile=$GISTMGR_URLFILE

    [ -z "$urlfile" ] && echo "$0: No urlfile present, abort!" && exit 2

    urlfolder=$(dirname $urlfile)
    if [ "$1" = "list" ]
    then
	if [ $# -ne 2 ]
	then
	    echo "$0: invalid number of arguments (got $#, need 2)" >&2
	    echo "$0: Usage: $0 list <list>" >&2
	    exit 3
	fi

	exec grep -e "^$2 " $urlfile
    elif [ $# -lt 3 ]
    then
	if [ $# -gt 1 ]
	then
	    echo "$0: invalid number of arguments (got $#, need 3)" >&2
	    echo "$0: Usage: $0 $1 <list> <url>" >&2
	else
	    echo "$0: command not found" >&2
	    usage
	fi
    elif [ "$1" = "add" ]
    then
	echo "$2 $(echo $3 | sed 's/ /%20/g') $4" >> $urlfile
	cd $urlfolder
	exec git commit -avm "[gistmgr] Add $3 to $2" | sed '1q'
    elif [ "$1" = "remove" ]
    then
	grep -ve "^$2 $(echo $3 | sed 's/ /%20/g').*$" $urlfile >/tmp/gistmgr-$(whoami)
	mv /tmp/gistmgr-$(whoami) $urlfile
	cd $urlfolder
	exec git commit -avm "[gistmgr] Remove $3 from $2" 2>&1 | sed '1q'
    fi
fi
